lib/GADS.pm

```diff
@@ -3028,6 +3030,69 @@ prefix '/:layout_name' => sub {
 
     };
 
+    any [ 'get', 'post' ] => '/historic_purge/' => require_login sub {
+        my $layout = var('layout') or pass;
+
+        forwardHome({ danger => 'You do not have permission to access this page.' })
+            unless $layout->user_can('purge');
+
+        my $user = logged_in_user;
+        my $view = current_view($user, $layout);
+
+        my %params = (
+            user   => $user,
+            layout => $layout,
+            schema => schema,
+            view   => $view,
+        );
+
+        if (defined param('stage1'))
+        {
+            $params{columns} = [ body_parameters->get_all('column_id') ];
+
+            my $records = GADS::Records->new(%params);
+            my @columns = grep { !$_->internal } @{ $records->columns_render };
+
+            my $table_data = [];
+            my $columns = {};
+
+            while (my $record = $records->single)
+            {
+                my @mapped_columns =$record->presentation_map_columns(columns => \@columns);
+                
+                my $values = {};
+
+                foreach my $column (@mapped_columns)
+                {
+                    $columns->{$column->{id}} = $column->{name};
+                    $values->{$column->{id}} = $column->{data}->{value} || "No value";
+                }
+
+                push @$table_data, {
+                    current_id => $record->current_id,
+                    values  => $values
+                };
+            }
+
+            return template "historic_purge/confirm" => { table_data => $table_data, columns => $columns };
+        }
+        elsif (defined param('purge'))
+        {
+            my $current_ids = [body_parameters->get_all('current_id')];
+            my $layouts     = [body_parameters->get_all('layout')];
+
+            if ( process( sub { schema->resultset('Current')->historic_purge($user, $current_ids, $layouts) } ) )
+            {
+                return forwardHome({ success => "Records have now been purged" }, $layout->identifier . '/data');
+            }
+        }
+
+        my $records = GADS::Records->new(%params);
+        my @columns = grep { !$_->internal } @{ $records->columns_render };
```

Comment:
This is repeated above in the first condition block, maybe move this one to the top and reuse?